### PR TITLE
[SPARK-44863][UI] Add a button to download thread dump as a txt in Spark UI

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR added a button to download thread dump as a txt w.r.t. jstack formatting

### Why are the changes needed?

The formatting of raw jstack can be relatively easy to read and analyze using various thread tools.

### Does this PR introduce _any_ user-facing change?

image

#### Raw Dump File
[driver.txt](https://github.com/apache/spark/files/12388302/driver.txt)

#### Reporting by external tools

https://fastthread.io/my-thread-report.jsp?p=c2hhcmVkLzIwMjMvMDgvMjAvZHJpdmVyLnR4dC0tMTMtNTMtMjI=&

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @gengliangwang @gatorsmile

**Comment by HyukjinKwon** [comment]
also @sarutak and @jasonli-db

File: core/src/main/scala/org/apache/spark/ui/exec/ExecutorThreadDumpPage.scala [hunk 1]
```diff
@@ -67,18 +69,17 @@ private[ui] class ExecutorThreadDumpPage(
...
-          <div class="form-inline">
-            <div class="bs-example" data-example-id="simple-form-inline">
-              <div class="form-group">
-                <div class="input-group">
-                  <label class="mr-2" for="search">Search:</label>
-                  <input type="text" class="form-control" id="search" oninput="onSearchStringChange()"></input>
+          <div style="display: flex; align-items: center;">
+            <a class="expandbutton" onClick="expandAllThreadStackTrace(true)">Expand All</a>
+            <a class="expandbutton d-none" onClick="collapseAllThreadStackTrace(true)">Collapse All</a>
+            <a class="downloadbutton" href={"data:text/plain;charset=utf-8," + threadDump.map(_.toString).mkString} download={"threaddump_" + executorId + ".txt"}>Download</a>
```
Comment (pan3793) [inline core/src/main/scala/org/apache/spark/ui/exec/ExecutorThreadDumpPage.scala:75]: how about adding "application id" as part of the file name?

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/ui/exec/ExecutorThreadDumpPage.scala:75]: When it comes to distinguishing files, there are several options such as using app id, attempt id, timestamps, executor details, and more to name the files. However, I prefer to keep it simple to allow for easy renaming when users click the button.

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala [hunk 2]
```diff
@@ -527,13 +527,53 @@ case class StackTrace(elems: Seq[String]) {
...
+    lockName: Option[String],
+    lockOwnerName: Option[String],
+    suspended: Boolean,
+    inNative: Boolean) {
+
+  /**
+   * Returns a string representation of this thread stack trace
+   * w.r.t java.lang.management.ThreadInfo(JDK 8)'s toString.
+   *
+   * TODO(Kent Yao): Considering 'daemon', 'priority' from higher JDKs
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/status/api/v1/api.scala:549]: should format the TODO as `TODO(SPARK-XXX):`

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala [hunk 3]
```diff
@@ -527,13 +527,53 @@ case class StackTrace(elems: Seq[String]) {
...
+   *
+   * TODO(Kent Yao): Also considering adding information os_prio, cpu, elapsed, tid, nid, etc.,
+   *   from the jstack tool
+   */
+  override def toString: String = {
+    val sb = new StringBuilder()
+    val basic = "\"" + threadName + "\" Id=" + threadId + " "  + threadState
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/status/api/v1/api.scala:556]:
```
val basic = s""""$threadName" Id=$threadId $threadState"""
sb.append(basic)
```
or

```
sb.append("\"").append(threadName).append("\" Id=").append(threadId).append(" ").append(threadState)
// sb.append("\"", threadName, "\" Id=", threadId, " ", threadState) multiple parameters api needs to wait until only Scala 2.13+ is supported.
```
The former has better readability, while the latter can avoid creating the intermediate string `basic`.

Comment (yaooqinn) [reply core/src/main/scala/org/apache/spark/status/api/v1/api.scala:556]: I choose another simple way `val sb = new StringBuilder(s""""$threadName" Id=$threadId $threadState""")`

**Comment by zhengruifeng** [comment]
cc @jasonli-db and @gengliangwang  would you mind take a look?

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Review by jasonli-db — APPROVED** [review]
LGTM!

File: project/MimaExcludes.scala [hunk 4]
```diff
@@ -36,6 +36,9 @@ object MimaExcludes {
 
   // Exclude rules for 4.0.x
   lazy val v40excludes = v35excludes ++ Seq(
+    // [SPARK-44863][UI] Add a button to download thread dump as a txt in Spark UI
```
Comment (zhengruifeng) [inline project/MimaExcludes.scala:39]: @yaooqinn  would you mind move the two exclusions to `v35excludes`, so that we can also merge it to 3.5?

Comment (yaooqinn) [reply project/MimaExcludes.scala]: OK, thanks, done

**Review by zhengruifeng — APPROVED** [review]

**Comment by yaooqinn** [comment]
thank you all, merged to master
https://github.com/apache/spark/pull/42654 is for branch 3.5

**Comment by zhengruifeng** [comment]
@yaooqinn thank you so much for working on this!

**Pull request closed without merging**